### PR TITLE
support batch request

### PR DESCRIPTION
**Pull request by qiweiii**
closes: #444 

image

### CODE REVIEW

**Comment by qiweiii** [comment]
Do we need to batch response?

**Comment by xlc** [comment]
yes. it should do the same thing as a normal Substrate client

**Comment by xlc** [comment]
and should have some tests

**Review by xlc — APPROVED** [review]

File: packages/chopsticks/src/server.ts [hunk 1]
```diff
@@ -126,36 +164,18 @@ export const createServer = async (handler: Handler, port?: number) => {
...
-        send({
-          id: req.id,
-          jsonrpc: '2.0',
-          error: e instanceof ResponseError ? e : { code: -32603, message: `Internal ${e}` },
-        })
+      if (Array.isArray(req)) {
+        logger.trace({ req }, 'Received batch request')
+        const resp: Awaited<ReturnType<typeof processRequest>>[] = []
+        for (const r of req) {
+          const result = await processRequest(r)
```
Comment (ermalkaleci) [inline packages/chopsticks/src/server.ts:171]: Make this Promise.all([req.map(processRequset)])

File: packages/chopsticks/src/server.ts [hunk 2]
```diff
@@ -126,36 +164,14 @@ export const createServer = async (handler: Handler, port?: number) => {
...
-      } catch (e) {
-        logger.info('Error handling request: %s %o', e, (e as Error).stack)
-        send({
-          id: req.id,
-          jsonrpc: '2.0',
-          error: e instanceof ResponseError ? e : { code: -32603, message: `Internal ${e}` },
-        })
+      if (Array.isArray(req)) {
+        logger.trace({ req }, 'Received batch request')
+        const resp = await Promise.all([...req.map(processRequest)])
```
Comment (ermalkaleci) [inline packages/chopsticks/src/server.ts:169]:
```suggestion
        const resp = await Promise.all(req.map(processRequest))
```
Small change

**Review by ermalkaleci — APPROVED** [review]

File: packages/testing/src/index.ts [hunk 3]
```diff
@@ -81,10 +82,21 @@ export const setupContextWithConfig = async ({ timeout, ...config }: SetupConfig
...
+  })
+  await isWsClientReady
+
   return {
     url,
     chain,
+    /** @polkadot/api WsProvider */
     ws,
+    /** WebSocket Client */
+    wsClient,
```
Comment (ermalkaleci) [inline packages/testing/src/index.ts:99]: do we need this?

Comment (qiweiii) [reply packages/testing/src/index.ts:99]: not sure if there are other ways to test this in e2e test

File: packages/e2e/src/batch-request.test.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import { afterAll, describe, expect, it } from 'vitest'
+
+import networks from './networks'
+
+describe('Batch request', async () => {
+  const acala = await networks.acala()
+  const { chain, wsClient } = acala
```
Comment (ermalkaleci) [inline packages/e2e/src/batch-request.test.ts:7]:
```suggestion
  const { chain, url } = acala
  const webSocket = new WebSocket(url)
```

maybe we can do this?

Comment (qiweiii) [reply packages/e2e/src/batch-request.test.ts:7]: yea, we can do this, but I feel it's cleaner to set it up in testing, and we may reuse it in future

Comment (ermalkaleci) [reply packages/e2e/src/batch-request.test.ts:7]: if we keep then it needs a proper error handling. Current code will not resolve if connect fails. So better remove it

Comment (qiweiii) [reply packages/e2e/src/batch-request.test.ts:7]: yes, we need to handle connect fail, but we still need to connect right? no matter where we put this code

Comment (ermalkaleci) [reply packages/e2e/src/batch-request.test.ts:7]: yes, put it on the batch test

**Pull request merged**